Sources/SwiftMemcache/MemcachedError.swift

```diff
@@ -0,0 +1,235 @@
+            self.storage.location = newValue
+        }
+    }
+
+    public init(
+        code: Code,
+        message: String,
+        cause: Error?,
+        location: SourceLocation
+    ) {
+        self.storage = Storage(code: code, message: message, cause: cause, location: location)
+    }
+
+    /// Creates a ``MemcachedError`` by wrapping the given `cause` and its location and code.
+    internal init(message: String, wrapping cause: MemcachedError) {
+        self.init(code: cause.code, message: message, cause: cause, location: cause.location)
+    }
+}
+
+extension MemcachedError: CustomStringConvertible {
+    public var description: String {
+        if let cause = self.cause {
+            return "\(self.code): \(self.message) (\(cause))"
+        } else {
+            return "\(self.code): \(self.message)"
+        }
+    }
+}
+
+extension MemcachedError: CustomDebugStringConvertible {
+    public var debugDescription: String {
+        if let cause = self.cause {
+            return "\(String(reflecting: self.code)): \(String(reflecting: self.message)) (\(String(reflecting: cause)))"
+        } else {
+            return "\(String(reflecting: self.code)): \(String(reflecting: self.message))"
+        }
+    }
+}
+
+extension MemcachedError {
+    private func detailedDescriptionLines() -> [String] {
+        // Build up a tree-like description of the error. This allows nested causes to be formatted
+        // correctly, especially when they are also MemcachedError.
+        var lines = [
+            "MemcachedError: \(self.code)",
+            "├─ Reason: \(self.message)",
+        ]
+
+        if let error = self.cause as? MemcachedError {
+            lines.append("├─ Cause:")
+            let causeLines = error.detailedDescriptionLines()
+            // We know this will never be empty.
+            lines.append("│  └─ \(causeLines.first!)")
+            lines.append(contentsOf: causeLines.dropFirst().map { "│     \($0)" })
+        } else if let error = self.cause {
+            lines.append("├─ Cause: \(String(reflecting: error))")
+        }
+
+        lines.append("└─ Source location: \(self.location.function) (\(self.location.file):\(self.location.line))")
+
+        return lines
+    }
+
+    /// A detailed multi-line description of the error.
+    ///
+    /// - Returns: A multi-line description of the error.
+    public func detailedDescription() -> String {
+        return self.detailedDescriptionLines().joined(separator: "\n")
+    }
+}
+
+extension MemcachedError {
+    /// A high level indication of the kind of error being thrown.
+    public struct Code: Hashable, Sendable, CustomStringConvertible {
+        private enum Wrapped: Hashable, Sendable, CustomStringConvertible {
+            /// Indicates that the connection has shut down.
+            case connectionShutdown
+            /// Indicates that a nil response was received from the server.
+            case unexpectedNilResponse
```

Comment:
I don't think we should expose this publicly and we should double check when we actually throw this. This really should only happen when the `inboundStream` returns us `nil`. In that case we should just throw `connectionShutdown`